### PR TITLE
Add Korean Revised Romanization to hangeul IME

### PR DESCRIPTION
**Pull request by annedrewhu**
Any Korean transliteration keyboard will have a lot of edge cases to check because of differences in romanization between revised romanization, the older McCune–Reischauer, and even a different system included in the RR standard. I've added a modest amount of tests based on this [official RR document](https://www.korean.go.kr/front_eng/roman/roman_01.do), but I'm open to adding more before merging.

### CODE REVIEW

File: rules/kor/kor-rr.js [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+		[ 'ᄃd', 'ᄄ' ],
+		[ 'ᄐd', 'ᄄ' ],
+		[ 'd', 'ᄃ' ],
+		[ 'r', 'ᄅ' ],
+		[ 'l', 'ᄅ' ],
+		[ 'm', 'ᄆ' ],
+		[ 'b', 'ᄇ' ],
+		[ 'ᄇp', 'ᄈ' ],
+		[ 'ᄉs', 'ᄊ' ],
+		// [ '\'', 'ᄋ'],  // Apostrophe can be written to represent silent ᄋ
```
Comment (amire80) [inline rules/kor/kor-rr.js:54]: Looks unnecessary.

File: rules/kor/kor-rr.js [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+		version: '1.0',
+		maxKeyLength: 4,
+		contextLength: 1,
+
+		// This function mirrors the normal behavior in jquery.ime.js, 
+		// except it combines jamo when a new syllable starts
+		// This version does not support context rules, but we don't need them
+		patterns: function(input, context) {
+			var patterns, regex, rule, replacement, i, result;
+
```
Comment (amire80) [inline rules/kor/kor-rr.js:125]: Too many empty lines. There should be one.

File: rules/kor/kor-rr.js [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+				// method can have a function as the second argument.
+				replacement = rule.slice( -1 )[ 0 ];
+
+				// Input string match test
+				if ( regex.test( input ) ) {
+					result = input.replace(regex, replacement);
+					
+					// This regex matches jamo that form a syllable so they can be combined
+					var jamoRegex = /([ᄀ-ᄒ])([ᅡ-ᅵ])([ᆨ-ᇂ])?([ᄀ-ᄒ]|[\- '])(.*)$/;
+					if (jamoRegex.test(result)) {
```
Comment (amire80) [inline rules/kor/kor-rr.js:143]: We require spaces inside parentheses.

File: rules/kor/kor-rr.js [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+						return { noop: false, output: result.replace(jamoRegex, combineJamo) };
```
Comment (amire80) [inline rules/kor/kor-rr.js:144]: Here, too, spaces inside parentheses.

File: rules/kor/kor-rr.js [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+			// No matches, return the input
+			return { noop: true, output: input };
+		},
+	};
+
+	// Conjoining jamo behavior is defined by this Unicode standard
+	// https://www.unicode.org/versions/Unicode13.0.0/ch03.pdf#G24646
+	// parameter `final` is optional
+	function combineJamo(substring, initial, vowel, final, nextSyllableInitial, otherChars) {
```
Comment (amire80) [inline rules/kor/kor-rr.js:159]: Here, too, spaces inside parentheses.

File: rules/kor/kor-rr.js [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+		var initialDiff = initialNo - 'ᄀ'.charCodeAt(0);
+		var vowelDiff = vowelNo - 'ᅡ'.charCodeAt(0);
+
+		// See Unicode standard: https://www.unicode.org/versions/Unicode13.0.0/ch03.pdf#G24646
+		var syllableNo = 44032 + initialDiff * 588 + vowelDiff * 28 + finalDiff;
+
+		var syllable = String.fromCharCode(syllableNo);
+
+		const disambig = /[\- ']/;
+		if (nextSyllableInitial.match(disambig)) {
```
Comment (amire80) [inline rules/kor/kor-rr.js:179]: Here, too, spaces inside parentheses.

File: rules/kor/kor-rr.js [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+			return syllable;
+		} else if (otherChars.match(disambig)) {
```
Comment (amire80) [inline rules/kor/kor-rr.js:181]: And here.

File: src/jquery.ime.inputmethods.js [hunk 8]
```diff
@@ -1274,6 +1278,10 @@
 			autonym: 'ಕನ್ನಡ',
 			inputmethods: [ 'kn-transliteration', 'kn-inscript', 'kn-kgp', 'kn-inscript2' ]
 		},
+		kor: {
+			autonym: '한국어',
+			inputmethods: [ 'kor-rr' ]
```
Comment (amire80) [inline src/jquery.ime.inputmethods.js:1283]: We usually give them an identified that is based on the shortest language code, so it should be "ko" and not "kor".

File: test/jquery.ime.test.fixtures.js [hunk 9]
```diff
@@ -3661,6 +3661,79 @@ var palochkaVariants = {
 			{ input: 'd~ha', output: 'ದ್ಹ', description: 'd~ha for ದ್ಹ in Kannada transliteration' }
 		]
 	},
+	{
+		description: 'Korean RR test',
+		inputmethod: 'kor-rr',
+		tests: [
+			// Note that RR is meant to romanize from hangul to latin script, but not
```
Comment (amire80) [inline test/jquery.ime.test.fixtures.js:3668]: Capitalize "Hangul" and "Latin".

File: rules/kor/kor-rr.js [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+		[ '([^ᄀ-ᄒ]|^)e', '$1에' ],
+		[ 'i', 'ᅵ' ],
+		[ 'a', 'ᅡ' ],
+		[ 'u', 'ᅮ' ],
+		[ 'o', 'ᅩ' ],
+		[ 'e', 'ᅦ' ],
+	];
+
+	var koreanRR = {
+		id: 'kor-rr',
```
Comment (amire80) [inline rules/kor/kor-rr.js:109]: We usually give them an identified that is based on the shortest language code, so it should be "ko" and not "kor".

**Review by amire80 — COMMENTED** [review]
Just a few silly, nitpicking comments about code style for now. We'll go over the code later.

**Review by srish — APPROVED** [review]
I tested this patch and it works as intended. @annedrewhu Thanks a lot for your contributions and apologies for the delay in reviewing!
